ecllibrary/std/system/Store.ecl

```diff
@@ -0,0 +1,573 @@
+                        STRING  pKey            {XPATH('Key')} := keyName;
+                        STRING  pValue          {XPATH('Value')} := keyValue;
+                        BOOLEAN pUserSpecific   {XPATH('UserSpecific')} := isUserSpecific;
+                    },
+                    DATASET(SetKeyValueResponseRec),
+                    XPATH('SetResponse'),
+                    HTTPHEADER('Authorization', ENCODED_CREDENTIALS),
+                    TIMEOUT(timeoutInSeconds)
+                );
+
+            RETURN UpdateForExceptions(soapResponse)[1];
+        END;
+
+        /**
+         * Gets a previously-set value for a key within a namespace.
+         *
+         * @param   keyName             A STRING naming the key; may not be an
+         *                              empty string; REQUIRED
+         * @param   isUserSpecific      If TRUE, the system will look only for
+         *                              private keys; if FALSE then the system will
+         *                              look for global keys; OPTIONAL, defaults
+         *                              to FALSE
+         * @param   timeoutInSeconds    The number of seconds to wait for the
+         *                              underlying SOAPCALL to complete; set to zero
+         *                              to wait forever; OPTIONAL, defaults to zero
+         *
+         * @return  A GetKeyValueResponseRec RECORD.  Note that the record will have
+         *          was_found set to TRUE or FALSE, depending on whether the key
+         *          was actually found in the key/value store.
+         */
+        EXPORT GetKeyValue(STRING keyName,
+                           BOOLEAN isUserSpecific = FALSE,
+                           UNSIGNED2 timeoutInSeconds = 0) := FUNCTION
+            soapResponse := SOAPCALL
+                (
+                    MY_ESP_URL,
+                    'Fetch',
+                    {
+                        STRING  pStoreName      {XPATH('StoreName')} := storeName;
+                        STRING  pNamespace      {XPATH('Namespace')} := namespace;
+                        STRING  pKey            {XPATH('Key')} := keyName;
+                        BOOLEAN pUserSpecific   {XPATH('UserSpecific')} := isUserSpecific;
+                    },
+                    DATASET(GetKeyValueResponseRec),
+                    XPATH('FetchResponse'),
+                    HTTPHEADER('Authorization', ENCODED_CREDENTIALS),
+                    TIMEOUT(timeoutInSeconds)
+                );
+
+            wasFound := (soapResponse[1].exceptions[1].exceptions[1].code != '-1');
+
+            finalResponse := IF
+                (
+                    wasFound,
+                    PROJECT
+                        (
+                            soapResponse,
+                            TRANSFORM
+                                (
+                                    RECORDOF(LEFT),
+                                    SELF.was_found := TRUE,
```

Comment:
Is it possible to have multiple responses from this service?  Why the conditional code to clear the record.

It would be much more efficient to use an unconditional project with 
   SELF.was_found := LEFT.exceptions[1].exceptions[1].code != '-1');

Reply:
The intention here is to clean up the structure returned to the caller.  Basically, if the call fails for any reason, the returned record just says "was_found = FALSE" and all other information was removed.  This is perhaps not the best solution anyway, because some failures are more important than others (e.g. an unknown namespace rather than an unknown key).  I'll look at cleaning this up.